From 3d9887a257149689d6a4b7ad446dc0dffdcd3080 Mon Sep 17 00:00:00 2001
From: Fabian Kosmale <fabian.kosmale@qt.io>
Date: Thu, 9 Oct 2025 11:46:01 +0200
Subject: [PATCH 169/239] CanvasImageData: Track image data size for gc
 purposes

Images can consume a substantial amount of memory on the native heap,
while the actual CanvasImageData object is rather small. Our gc already
keeps track of native heap consumption (e.g. for strings), so reuse that
support to also properly account for CanvasImageData.

Pick-to: 6.8
Fixes: QTBUG-32298
Change-Id: Ie36e9c2399517154d9c46e490a49b16409cbecc1
Reviewed-by: Ulf Hermann <ulf.hermann@qt.io>
(cherry picked from commit eee4e767ab77e6643b0b5992a6879b2f8230f700)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quick/items/context2d/qquickcontext2d.cpp  |  7 ++++++-
 .../data/tst_grabManyImages.qml                | 18 ++++++++++++++++++
 2 files changed, 24 insertions(+), 1 deletion(-)
 create mode 100644 tests/auto/quick/qquickcanvasitem/data/tst_grabManyImages.qml

diff --git a/src/quick/items/context2d/qquickcontext2d.cpp b/src/quick/items/context2d/qquickcontext2d.cpp
index 7102ab27f4..ed283213b5 100644
--- a/src/quick/items/context2d/qquickcontext2d.cpp
+++ b/src/quick/items/context2d/qquickcontext2d.cpp
@@ -497,7 +497,11 @@ struct QQuickContext2DStyle : Object {
 struct QQuickJSContext2DPixelData : Object {
     void init();
     void destroy() {
-        delete image;
+        if (image) {
+            auto *mm = internalClass->engine->memoryManager;
+            mm->changeUnmanagedHeapSizeUsage(-image->sizeInBytes());
+            delete image;
+        }
         Object::destroy();
     }
 
@@ -940,6 +944,7 @@ static QV4::ReturnedValue qt_create_image_data(qreal w, qreal h, QV4::ExecutionE
     QV4::Scope scope(v4);
     QQuickContext2DEngineData *ed = engineData(scope.engine);
     QV4::Scoped<QQuickJSContext2DPixelData> pixelData(scope, scope.engine->memoryManager->allocate<QQuickJSContext2DPixelData>());
+    v4->memoryManager->changeUnmanagedHeapSizeUsage(image.sizeInBytes());
     QV4::ScopedObject p(scope, ed->pixelArrayProto.value());
     pixelData->setPrototypeOf(p);
 
diff --git a/tests/auto/quick/qquickcanvasitem/data/tst_grabManyImages.qml b/tests/auto/quick/qquickcanvasitem/data/tst_grabManyImages.qml
new file mode 100644
index 0000000000..fd18099866
--- /dev/null
+++ b/tests/auto/quick/qquickcanvasitem/data/tst_grabManyImages.qml
@@ -0,0 +1,18 @@
+import QtQuick 2.0
+
+Item {
+    width: 800
+    height: 800
+
+    Canvas {
+        id: canvas
+        renderStrategy: Canvas.Immediate
+        anchors.fill: parent
+        onPaint: {}
+        onAvailableChanged: {
+            // must not run out of memory
+            for (let i = 0; i < 2000; ++i)
+                canvas.getContext("2d").getImageData(0, 0, canvas.width, canvas.height);
+        }
+    }
+}
-- 
2.51.2

